### PR TITLE
Fix #10664 by restricting generalize_structure

### LINKED ISSUE #10664: Polymorphic methods and fields on polymorphic variants
There seems to be a problem with compiling pattern-matches against the application of a polymorphic method or field to a polymorphic variant.  Here is an example with a method:
```ocaml
class idfunc =
  object
    method id : 'ab. ([< `A | `B ] as 'ab) -> 'ab = fun x -> x
  end

let act : [ `A | `B ] -> string =
 fun x -> match (new idfunc)#id x with `B -> "B" | `A -> "A"

let _ = print_endline (act `A)
```
This prints `B`, whereas it should print `A`.  Here is a similar example with a field:
```ocaml
type idfunc = { id : 'ab. ([< `A | `B ] as 'ab) -> 'ab }

let f = { id = (fun x -> x) }

let act : [ `A | `B ] -> string =
 fun x -> match f.id x with `B -> "B" | `A -> "A"

let _ = print_endline (act `A)
```
This also prints `B`.

The problem goes away if any of the following changes are made:

- Changing the type of `id` to `'ab. 'ab -> 'ab`.
- Changing the type of `id` to ``[ `A | `B ] -> [ `A | `B ]``.
- Coercing the value `(new idfunc)#id x` or `f.id x` to ``[ `A | `B ]`` before matching.

In addition,

- Calling ``act `B`` instead *also* prints `B`.
- Switching the order of the match clauses causes the code to always print `A` instead.
- Adding a catchall case ``| _ -> "C"`` at the end of the match (with or without changing the type of `act` to ``[> `A | `B ] -> string``) changes the return value to `C`.

Some brief initial discussion was [here](https://discuss.ocaml.org/t/polymorphic-methods-on-polymorphic-variants/8533?u=viritrilbia).

### PR DESCRIPTION
**Pull request by garrigue**
@gasche correctly analyzed that the problem in #10664 is related to calling `generalize_structure` inside pattern-matching.

More precisely, using `generalize_structure` on non-static polymorphic variants may result in losing the necessary sharing of the including `row_field_desc`.

This PR fixes that problem by just not generalizing non-static polymorphic variants when they are not fixed (when they are fixed, they cannot be modified in place, so the loss of sharing is not a problem).
To do so, it calls `update_level` without environment inside `generalize_structure`. Since the call to `generalize_structure` is supposed to happen just after `end_def`, no scoping is involved, so we assume that no error can occur, hence the `assert false` in case of exception.

### CODE REVIEW

**Comment by garrigue** [comment]
Looking more carefully at the test suite, this changes can break programs that match a GADT constructor under a polymorphic variant. This has always been fragile, so I think this is OK; are there really people using that kind of code?

By the way, this can still be fixed by adding an annotation that makes the polymorphic variant static.

**Comment by yakobowski** [comment]
> This has always been fragile, so I think this is OK; are there really people using that kind of code?

I will need to check, but I think we do in some cases

> By the way, this can still be fixed by adding an annotation that makes the polymorphic variant static.

Completely monomorphic (e.g. for the type of the arguments of the constructors), or just without a row variable?

**Comment by garrigue** [comment]
> > By the way, this can still be fixed by adding an annotation that makes the polymorphic variant static.
> 
> Completely monomorphic (e.g. for the type of the arguments of the constructors), or just without a row variable?

Just without a row variable. The example in `or_patterns.ml` shows the problem, which is solved just after by the annotated version. The annotated could even contain free variables, as long as they do not make existential variables leak.

**Comment by garrigue** [comment]
To clarify, this is a soundness bug.
I'm waiting for a review so that we can close it.

Note that `generalize_structure` is about principality under minimal derivations, so that this changes does not break its properties. It actually makes easier to reason about it.
I still need to understand all the implications on principality (generalize_structure is anyway more about minimal derivations than principality), but soundness trumps strict principality.

File: typing/ctype.ml [hunk 1]
```diff
@@ -618,6 +620,8 @@ let rec generalize_structure ty =
       match get_desc ty with
         Tconstr (p, _, abbrev) ->
           not (is_object_type p) && (abbrev := Mnil; true)
+      | Tvariant row when not (is_fixed row) && not (static_row row) ->
```
Comment (lpw25) [inline typing/ctype.ml:623]: I think `row_fixed row` rather than `is_fixed row` would be fine and allow more cases to type-check.

**Review by lpw25 — APPROVED** [review]
The problem and the fix make sense to me. It's a shame to lose the ability to know principal information withing polymorphic variants, but it seems unavoidable given the invariants around sharing of polymorphic variants.

Are those invariants written down somewhere? They would make it easier to think about issues like this.

One thing that surprises me a little is that there doesn't seem to be any equivalent code for objects. I thought they had similar sharing invariants, and I would expect them to also be broken by `generalize_structure`.

**Comment by garrigue** [comment]
@lpw25 Thanks for looking at that !
However, your comments led me a to a more in depth analysis of the problem, resulting in #11600, which replaces this PR.
Still need to understand why, and under which conditions, `generalize_structure` is OK.
The ultimate solution is probably to switch to a representation of rows more in line with the theory, but this is going to break some programs.

**Comment by garrigue** [comment]
Subsumed by #11600

**Pull request closed without merging**